File: src/main/sync/git.ts
import { spawn } from 'node:child_process'
import { basename } from 'node:path'

// Thin promisified git/gh runner for sync. Follows the extract/claude.ts spawn
// discipline: spawn can throw SYNCHRONOUSLY under fd pressure (EBADF/EMFILE),
// so every call is guarded and resolves a result instead of ever rejecting.

export interface RunResult {
  code: number
  stdout: string
  stderr: string
}

const LOCAL_TIMEOUT_MS = 30_000
const NETWORK_TIMEOUT_MS = 120_000

export function run(bin: string, args: string[], opts: { cwd?: string; env?: Record<string, string>; timeoutMs?: number } = {}): Promise<RunResult> {
  return new Promise((resolve) => {
    let stdout = ''
    let stderr = ''
    let child
    try {
      child = spawn(bin, args, {
        cwd: opts.cwd,
        // GIT_TERMINAL_PROMPT=0: a missing credential fails fast instead of
        // wedging the sync on an invisible username prompt.
        env: { ...process.env, GIT_TERMINAL_PROMPT: '0', ...opts.env },
        stdio: ['ignore', 'pipe', 'pipe']
      })
    } catch (e) {
      resolve({ code: -1, stdout: '', stderr: String(e) })
      return
    }
    const killer = setTimeout(() => child.kill('SIGKILL'), opts.timeoutMs ?? LOCAL_TIMEOUT_MS)
    child.stdout?.on('data', (d) => (stdout += d))
    child.stderr?.on('data', (d) => (stderr += d))
    child.on('error', (e) => {
      clearTimeout(killer)
      resolve({ code: -1, stdout, stderr: stderr || String(e) })
    })
    child.on('close', (code) => {
      clearTimeout(killer)
      resolve({ code: code ?? -1, stdout, stderr })
    })
  })
}

export type GitAuth =
  | { mode: 'github-app'; token: string }
  | { mode: 'gh-cli' }
  | { mode: 'git' }

/** Credential wiring that never touches disk or argv-visible secrets: the
 *  token rides an env var read by an inline helper; gh mode delegates to gh's
 *  own credential store. The leading empty helper clears any system helper so
 *  a stale keychain credential can't shadow the intended one. */
function authArgs(auth: GitAuth): { args: string[]; env: Record<string, string> } {
  if (auth.mode === 'github-app') {
    return {
      args: [
        '-c', 'credential.helper=',
        '-c', 'credential.helper=!f() { echo username=x-access-token; echo "password=$ZEDE_SYNC_TOKEN"; }; f'
      ],
      env: { ZEDE_SYNC_TOKEN: auth.token }
    }
  }
  if (auth.mode === 'gh-cli') {
    return { args: ['-c', 'credential.helper=', '-c', 'credential.helper=!gh auth git-credential'], env: {} }
  }
  return { args: [], env: {} }
}

export const git = (dir: string, args: string[], opts: { auth?: GitAuth; timeoutMs?: number } = {}): Promise<RunResult> => {
  const a = opts.auth ? authArgs(opts.auth) : { args: [], env: {} }
  return run('git', ['-C', dir, ...a.args, ...args], { env: a.env, timeoutMs: opts.timeoutMs })
}

export async function gitAvailable(): Promise<boolean> {
  return (await run('git', ['--version'])).code === 0
}

export async function ghAvailable(): Promise<boolean> {
  return (await run('gh', ['auth', 'status'])).code === 0
}

/** Init (idempotent) + point origin at the remote. Identity is passed per-commit
 *  via -c, so nothing is written to the user's git config. */
export async function ensureRepo(dir: string, remoteUrl: string): Promise<string | null> {
  const init = await git(dir, ['init', '-b', 'main'])
  if (init.code !== 0) return `git init failed: ${init.stderr.trim()}`
  const has = await git(dir, ['remote', 'get-url', 'origin'])
  const r = has.code === 0
    ? await git(dir, ['remote', 'set-url', 'origin', remoteUrl])
    : await git(dir, ['remote', 'add', 'origin', remoteUrl])
  return r.code === 0 ? null : `git remote failed: ${r.stderr.trim()}`
}

export async function fetchMain(dir: string, auth: GitAuth): Promise<'ok' | 'no-remote-branch' | 'offline'> {
  const r = await git(dir, ['fetch', 'origin', 'main'], { auth, timeoutMs: NETWORK_TIMEOUT_MS })
  if (r.code === 0) return 'ok'
  // A brand-new remote has no main yet — that's a fine state, not offline.
  if (/couldn't find remote ref/i.test(r.stderr)) return 'no-remote-branch'
  return 'offline'
}

/** Adopt the remote tree exactly. Guarded so a bug can never hard-reset anything
 *  but the dedicated sync working copy (the tree is regenerable from the DB). */
export async function resetToRemote(dir: string): Promise<boolean> {
  if (basename(dir) !== 'sync') throw new Error(`refusing to hard-reset non-sync dir: ${dir}`)
  return (await git(dir, ['reset', '--hard', 'origin/main'])).code === 0
}

/** Stage everything and commit if anything changed. Returns whether a commit was made. */
export async function commitAll(dir: string, message: string): Promise<{ committed: boolean; error?: string }> {
  const add = await git(dir, ['add', '-A'])
  if (add.code !== 0) return { committed: false, error: add.stderr.trim() }
  const clean = (await git(dir, ['diff', '--cached', '--quiet'])).code === 0
  const noHead = (await git(dir, ['rev-parse', '--verify', 'HEAD'])).code !== 0
  if (clean && !noHead) return { committed: false }
  const c = await git(dir, ['-c', 'user.name=Zede Sync', '-c', 'user.email=sync@zede.local', 'commit', '-m', message, '--allow-empty-message'])
  if (c.code !== 0) {
    // Nothing to commit on a fresh repo (empty DB export is never truly empty —
    // zede.json always exists — so this is defensive only).
    if (/nothing to commit/i.test(c.stdout + c.stderr)) return { committed: false }
    return { committed: false, error: c.stderr.trim() }
  }
  return { committed: true }
}

export async function push(dir: string, auth: GitAuth): Promise<'ok' | 'rejected' | 'auth' | 'offline'> {
  const r = await git(dir, ['push', '-u', 'origin', 'main'], { auth, timeoutMs: NETWORK_TIMEOUT_MS })
  if (r.code === 0) return 'ok'
  const err = r.stderr
  if (/\[rejected\]|non-fast-forward|fetch first/i.test(err)) return 'rejected'
  if (/authentication|permission|403|401|could not read Username|invalid credentials/i.test(err)) return 'auth'
  return 'offline'
}

// --- gh CLI helpers (fallback auth path) ---

export async function ghLogin(): Promise<string | null> {
  const r = await run('gh', ['api', 'user', '-q', '.login'], { timeoutMs: NETWORK_TIMEOUT_MS })
  return r.code === 0 ? r.stdout.trim() || null : null
}

/** Ensure `owner/name` exists (create private if missing). Returns clone URL or error. */
export async function ghEnsureRepo(name: string): Promise<{ url?: string; error?: string }> {
  const view = await run('gh', ['repo', 'view', name, '--json', 'url', '-q', '.url'], { timeoutMs: NETWORK_TIMEOUT_MS })
  if (view.code === 0 && view.stdout.trim()) return { url: `${view.stdout.trim()}.git` }
  const create = await run('gh', ['repo', 'create', name, '--private'], { timeoutMs: NETWORK_TIMEOUT_MS })
  if (create.code !== 0) return { error: create.stderr.trim() || 'gh repo create failed' }
  const view2 = await run('gh', ['repo', 'view', name, '--json', 'url', '-q', '.url'], { timeoutMs: NETWORK_TIMEOUT_MS })
  if (view2.code === 0 && view2.stdout.trim()) return { url: `${view2.stdout.trim()}.git` }
  return { error: 'created repo but could not resolve its URL' }
}

File: native/src/sync/git.rs
//! Thin git runner for sync, ported from `src/main/sync/git.ts`. Every call
//! resolves to a result instead of panicking; network calls get the long
//! timeout. `GIT_TERMINAL_PROMPT=0` makes a missing credential fail fast
//! instead of wedging on an invisible prompt.

use std::path::Path;
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};

const LOCAL_TIMEOUT: Duration = Duration::from_secs(30);
const NETWORK_TIMEOUT: Duration = Duration::from_secs(120);

pub struct RunResult {
    pub code: i32,
    pub stdout: String,
    pub stderr: String,
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum GitAuth {
    /// Plain git: whatever credentials the user's git already has (ssh keys,
    /// credential helpers, a NAS remote, a local bare repo…).
    Git,
    /// Delegate credentials to the GitHub CLI's own store.
    GhCli,
}

fn auth_args(auth: GitAuth) -> Vec<String> {
    match auth {
        GitAuth::Git => Vec::new(),
        // The leading empty helper clears any system helper so a stale
        // keychain credential can't shadow the intended one.
        GitAuth::GhCli => vec![
            "-c".into(),
            "credential.helper=".into(),
            "-c".into(),
            "credential.helper=!gh auth git-credential".into(),
        ],
    }
}

fn run(bin: &str, args: &[String], cwd: Option<&Path>, timeout: Duration) -> RunResult {
    let mut cmd = Command::new(bin);
    cmd.args(args)
        .env("GIT_TERMINAL_PROMPT", "0")
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    if let Some(dir) = cwd {
        cmd.current_dir(dir);
    }
    let mut child = match cmd.spawn() {
        Ok(c) => c,
        Err(e) => return RunResult { code: -1, stdout: String::new(), stderr: e.to_string() },
    };

    let stdout = child.stdout.take();
    let stderr = child.stderr.take();
    let out_handle = std::thread::spawn(move || {
        use std::io::Read;
        let mut s = String::new();
        if let Some(mut p) = stdout {
            let _ = p.read_to_string(&mut s);
        }
        s
    });
    let err_handle = std::thread::spawn(move || {
        use std::io::Read;
        let mut s = String::new();
        if let Some(mut p) = stderr {
            let _ = p.read_to_string(&mut s);
        }
        s
    });

    let start = Instant::now();
    let code = loop {
        match child.try_wait() {
            Ok(Some(status)) => break status.code().unwrap_or(-1),
            Ok(None) => {
                if start.elapsed() > timeout {
                    let _ = child.kill();
                    let _ = child.wait();
                    break -1;
                }
                std::thread::sleep(Duration::from_millis(100));
            }
            Err(_) => break -1,
        }
    };
    RunResult {
        code,
        stdout: out_handle.join().unwrap_or_default(),
        stderr: err_handle.join().unwrap_or_default(),
    }
}

pub fn git(dir: &Path, args: &[&str], auth: GitAuth, network: bool) -> RunResult {
    let mut full: Vec<String> = vec!["-C".into(), dir.to_string_lossy().into_owned()];
    full.extend(auth_args(auth));
    full.extend(args.iter().map(|s| s.to_string()));
    run("git", &full, None, if network { NETWORK_TIMEOUT } else { LOCAL_TIMEOUT })
}

pub fn git_available() -> bool {
    run("git", &["--version".into()], None, LOCAL_TIMEOUT).code == 0
}

/// Init (idempotent) + point origin at the remote. Identity is passed
/// per-commit via -c, so nothing is written to the user's git config.
pub fn ensure_repo(dir: &Path, remote_url: &str) -> Option<String> {
    let init = git(dir, &["init", "-b", "main"], GitAuth::Git, false);
    if init.code != 0 {
        return Some(format!("git init failed: {}", init.stderr.trim()));
    }
    let has = git(dir, &["remote", "get-url", "origin"], GitAuth::Git, false);
    let r = if has.code == 0 {
        git(dir, &["remote", "set-url", "origin", remote_url], GitAuth::Git, false)
    } else {
        git(dir, &["remote", "add", "origin", remote_url], GitAuth::Git, false)
    };
    if r.code == 0 {
        None
    } else {
        Some(format!("git remote failed: {}", r.stderr.trim()))
    }
}

pub enum FetchOutcome {
    Ok,
    NoRemoteBranch,
    Offline,
}

pub fn fetch_main(dir: &Path, auth: GitAuth) -> FetchOutcome {
    let r = git(dir, &["fetch", "origin", "main"], auth, true);
    if r.code == 0 {
        return FetchOutcome::Ok;
    }
    // A brand-new remote has no main yet — a fine state, not offline.
    if r.stderr.to_lowercase().contains("couldn't find remote ref") {
        return FetchOutcome::NoRemoteBranch;
    }
    FetchOutcome::Offline
}

/// Adopt the remote tree exactly. Guarded so a bug can never hard-reset
/// anything but the dedicated sync working copy (regenerable from the DB).
pub fn reset_to_remote(dir: &Path) -> Result<bool, String> {
    if dir.file_name().and_then(|n| n.to_str()) != Some("sync") {
        return Err(format!("refusing to hard-reset non-sync dir: {}", dir.display()));
    }
    Ok(git(dir, &["reset", "--hard", "origin/main"], GitAuth::Git, false).code == 0)
}

/// Stage everything and commit if anything changed.
pub fn commit_all(dir: &Path, message: &str) -> Result<bool, String> {
    let add = git(dir, &["add", "-A"], GitAuth::Git, false);
    if add.code != 0 {
        return Err(add.stderr.trim().to_string());
    }
    let clean = git(dir, &["diff", "--cached", "--quiet"], GitAuth::Git, false).code == 0;
    let no_head = git(dir, &["rev-parse", "--verify", "HEAD"], GitAuth::Git, false).code != 0;
    if clean && !no_head {
        return Ok(false);
    }
    let c = git(
        dir,
        &[
            "-c", "user.name=Zede Sync",
            "-c", "user.email=sync@zede.local",
            "commit", "-m", message, "--allow-empty-message",
        ],
        GitAuth::Git,
        false,
    );
    if c.code != 0 {
        let all = format!("{}{}", c.stdout, c.stderr).to_lowercase();
        if all.contains("nothing to commit") {
            return Ok(false);
        }
        return Err(c.stderr.trim().to_string());
    }
    Ok(true)
}

pub enum PushOutcome {
    Ok,
    Rejected,
    Auth,
    Offline,
}

pub fn push(dir: &Path, auth: GitAuth) -> PushOutcome {
    let r = git(dir, &["push", "-u", "origin", "main"], auth, true);
    if r.code == 0 {
        return PushOutcome::Ok;
    }
    let err = r.stderr.to_lowercase();
    if err.contains("[rejected]") || err.contains("non-fast-forward") || err.contains("fetch first")
    {
        return PushOutcome::Rejected;
    }
    if err.contains("authentication")
        || err.contains("permission")
        || err.contains("403")
        || err.contains("401")
        || err.contains("could not read username")
        || err.contains("invalid credentials")
    {
        return PushOutcome::Auth;
    }
    PushOutcome::Offline
}
